app/src/main/java/org/schabi/newpipe/player/PlayerService.kt

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2017 Mauricio Colli <mauriciocolli@outlook.com>
+ * Part of NewPipe
+ *
+ * License: GPL-3.0+
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.schabi.newpipe.player
+
+import android.content.Context
+import android.content.Intent
+import android.os.Binder
+import android.os.Bundle
+import android.os.IBinder
+import android.support.v4.media.MediaBrowserCompat
+import android.util.Log
+import androidx.media.MediaBrowserServiceCompat
+import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector
+import io.reactivex.rxjava3.disposables.CompositeDisposable
+import org.schabi.newpipe.player.mediabrowser.MediaBrowserConnector
+import org.schabi.newpipe.player.notification.NotificationPlayerUi
+import org.schabi.newpipe.util.Localization
+import org.schabi.newpipe.util.ThemeHelper
+import java.lang.ref.WeakReference
+
+/**
+ * One service for all players.
+ */
+class PlayerService : MediaBrowserServiceCompat() {
+    private val player: Player by lazy {
+        Player(this).apply {
+            /*
+            Create the player notification and start immediately the service in foreground,
+            otherwise if nothing is played or initializing the player and its components (especially
+            loading stream metadata) takes a lot of time, the app would crash on Android 8+ as the
+            service would never be put in the foreground while we said to the system we would do so
+             */
+            UIs()[NotificationPlayerUi::class.java]
+                .ifPresent { it.createNotificationAndStartForeground() }
+        }
+    }
+
+    private val mBinder: IBinder = LocalBinder(this)
+    private val compositeDisposableLoadChildren = CompositeDisposable()
+    private var mediaBrowserConnector: MediaBrowserConnector? = null
+        get() {
+            if (field == null) {
+                field = MediaBrowserConnector(this)
+            }
+            return field
+        }
```

Comment:
This avoids nullability checks down the line. Maybe this should be synchronized to avoid multiple threads from creating multiple media browser connectors? Also, why not just initialize this in `onCreate()`?
```suggestion
    private var _mediaBrowserConnector: MediaBrowserConnector? = null
    private val mediaBrowserConnector: MediaBrowserConnector
        get() {
            return _mediaBrowserConnector ?: run {
                val newMediaBrowserConnector = MediaBrowserConnector(this)
                _mediaBrowserConnector = newMediaBrowserConnector
                newMediaBrowserConnector
            }
        }
```

Reply:
This is still unresolved:
- `mediaBrowserConnector` is still followed by `?.` everywhere in the file, while it should be `.` after this change
- Now that I think of it, we can probably do: `private val mediaBrowserConnector by lazy { MediaBrowserConnector(this) }`
- Why not just initialize this in onCreate()? This would simplify everything, so please let us know if there is a reason why this can't be done there.